### ISSUE TITLE
Generated Insert command has wrong id of parent entity

### ISSUE THREAD

**Issue by sm-g** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#379]
### Steps to reproduce
Use https://github.com/sm-g/EFCoreInMemorySample
Create db using ([schema.sql](https://github.com/sm-g/EFCoreInMemorySample/blob/master/schema.sql)), 
update [connection string](https://github.com/sm-g/EFCoreInMemorySample/blob/master/WebApp/Startup.cs#L33)
run app, call http://localhost:5501/api/sampledata/test 
[This inserts](https://github.com/sm-g/EFCoreInMemorySample/blob/master/WebApp/Controllers/SampleDataController.cs#L21) will be executed. 

### The issue
After upgrade to EF Core 2 `SELECT ROW_COUNT();` affects following `LAST_INSERT_ID()`.

mysql log file:
```
2017-09-06T15:43:33.427420Z	   41 Query	start transaction
2017-09-06T15:43:33.521411Z	   41 Query	INSERT INTO `import_customer` (`customer_id`, `import_id`, `name`)
VALUES (1, 5, 'cust');
2017-09-06T15:43:33.521411Z	   41 Query	SELECT `id` FROM `import_customer` WHERE `id` = LAST_INSERT_ID();
2017-09-06T15:43:33.522411Z	   41 Query	INSERT INTO `import_group` (`group_id`, `import_id`, `name`)
VALUES (1, 5, 'cat1');
2017-09-06T15:43:33.522411Z	   41 Query	SELECT `id` FROM `import_group` WHERE `id` = LAST_INSERT_ID();
2017-09-06T15:43:33.522411Z	   41 Query	INSERT INTO `import_group` (`group_id`, `import_id`, `name`)
VALUES (2, 5, 'cat2');
2017-09-06T15:43:33.522411Z	   41 Query	SELECT `id` FROM `import_group` WHERE `id` = LAST_INSERT_ID();
2017-09-06T15:43:33.522411Z	   41 Query	INSERT INTO `import_summary` (`import_id`, `value`)
VALUES (5, true);
2017-09-06T15:43:33.573405Z	   41 Query	SELECT ROW_COUNT();
2017-09-06T15:43:33.574405Z	   41 Query	INSERT INTO `import_zone` (`import_id`, `name`, `zone_id`)
VALUES (5, 'zone1', 1);
2017-09-06T15:43:33.574405Z	   41 Query	SELECT `id` FROM `import_zone` WHERE `id` = LAST_INSERT_ID();
2017-09-06T15:43:33.574405Z	   41 Query	INSERT INTO `import_zone` (`import_id`, `name`, `zone_id`)
VALUES (5, 'zone2', 2);
2017-09-06T15:43:33.574405Z	   41 Query	SELECT `id` FROM `import_zone` WHERE `id` = LAST_INSERT_ID()
2017-09-06T15:43:33.579405Z	   41 Query	INSERT INTO `import_project` (`import_customer_id`, `import_id`, `name`, `project_id`)
VALUES (5, 5, 'proj1', 1);
2017-09-06T15:43:33.580405Z	   41 Query	SELECT `id` FROM `import_project` WHERE `id` = LAST_INSERT_ID();
2017-09-06T15:43:33.580405Z	   41 Query	INSERT INTO `import_project` (`import_customer_id`, `import_id`, `name`, `project_id`)
VALUES (5, 5, 'proj2', 2);
2017-09-06T15:43:33.580405Z	   41 Query	SELECT `id` FROM `import_project` WHERE `id` = LAST_INSERT_ID();
2017-09-06T15:43:33.580405Z	   41 Query	INSERT INTO `import_user` (`first_name`, `import_group_id`, `import_id`, `import_zone_id`, `last_name`, `middle_name`, `user_id`)
VALUES ('fn', 7, 5, 1, 'ln', '', 50);
```

The issue is gone if you remove inserting `import_summary` in same transaction with other inserts.

The first executing (when first id has 1 value), exception will be from EF change tracker, about "The instance of entity type ImportZone cannot be tracked because another instance with the key value 'Id:1' is already being tracked.". 

After this the issue becomes more clear, if you disable `fk__import_user__import_zone` to see what was inserted:

```
INSERT INTO `import_group` (`group_id`, `import_id`, `name`) ...
SELECT `id` FROM `import_group` WHERE `id` = LAST_INSERT_ID(); 
INSERT INTO `import_summary` (`import_id`,  ... → created with import_id 1
SELECT ROW_COUNT();
INSERT INTO `import_zone` (`import_id`, `name`, `zone_id`) ... → created with id 4
SELECT `id` FROM `import_zone` WHERE `id` = LAST_INSERT_ID();
INSERT INTO `import_zone` (`import_id`, `name`, `zone_id`)... → created with id 5
SELECT `id` FROM `import_zone` WHERE `id` = LAST_INSERT_ID(); 
INSERT INTO `import_zone` (`import_id`, `name`, `zone_id`)... → created with id 6
SELECT `id` FROM `import_zone` WHERE `id` = LAST_INSERT_ID(); 

`import_user`'s with first zone has `import_user_id == 1`, 
with second zone has `import_user_id == 4`
with third zone has `import_user_id == 5`
```

### Further technical details

MySQL version: 5.7.17
Operating system: win10
Pomelo.EntityFrameworkCore.MySql version: 2.0.0-rtm-10062

**Comment by caleblloyd** [comment]
If this only happens with your project and is not easily reproducible, it is likely a misconfiguration.

What is really best is if you can come up with a failing test case for our functional tests so that we can debug against the source:

https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/tree/master/test/EFCore.MySql.FunctionalTests

I recommend you look to make sure IDs and Foreign Keys are explicitly defined using the Fluent API to debug this on your side. Sometimes EF can can predict the wrong keys if annotations are used, Fluent API takes the guesswork out of it.

**Comment by sm-g** [comment]
As you see, no annotations used, everything by convention. EF model DebugView of Context is valid:
```
Foreign keys: 
      ImportUser {'ImportGroupId'} -> ImportGroup {'Id'} ToPrincipal: ImportGroup
      ImportUser {'ImportId'} -> Import {'Id'} ToPrincipal: Import
      ImportUser {'ImportZoneId'} -> ImportZone {'Id'} ToPrincipal: ImportZone
```
Same is when I explicitly set 
```
modelBuilder.Entity<ImportUser>()
               .HasOne(x => x.ImportZone)
               .WithMany()
               .HasForeignKey(x => x.ImportZoneId);
```

**Comment by sm-g** [comment]
@caleblloyd I update my first post, you can reproduce issue now.

**Comment by malmgren80** [comment]
I have the same issue. Tested with `2.0.0` and `2.0.1-preview-10066`

It seems like the insert after ROW_COUNT() ignores previous LAST_INSERT_ID() for FKs.

I don't know if ROW_COUNT() causes this issue. For me, it works when an UPDATE is issued instead of an INSERT prior to ROW_COUNT().

Mysql logs looks something like this (all Id columns are autoincremented):

Non working example:
```
INSERT INTO `Foo` (`Name`) VALUES ('Foo');
SELECT `Id` FROM `Foo` WHERE `Id` = LAST_INSERT_ID();

/* Another INSERT that causes ROW_COUNT() */

SELECT ROW_COUNT();
INSERT INTO `Bar` (`FooId`, `Name`) VALUES (28, 'Bar'); <-- Prev Id used correctly
SELECT `Id` FROM `Bar` WHERE `Id` = LAST_INSERT_ID();
INSERT INTO `Baz` (`BarId`, `Name`) VALUES (1, 'Baz');  <--- WRONG ID, becomes 1 instead of Bar.Id 
```

Working example:
```
INSERT INTO `Foo` (`Name`) VALUES ('Foo');
SELECT `Id` FROM `Foo` WHERE `Id` = LAST_INSERT_ID();

/* UPDATE that causes ROW_COUNT() */

SELECT ROW_COUNT();
INSERT INTO `Bar` (`FooId`, `Name`) VALUES (28, 'Bar'); <-- Prev Id used correctly
SELECT `Id` FROM `Bar` WHERE `Id` = LAST_INSERT_ID();
INSERT INTO `Baz` (`BarId`, `Name`) VALUES (54, 'Baz');  <-- Prev Id used correctly
```

**Comment by malmgren80** [comment]
After much digging I found a workaround by setting MaxBatchSize to 1 (as in `EFCore.MySql.FunctionalTests`):

```
options.UseMySql(
        connectionString,
        mysqlOptions => mysqlOptions.MaxBatchSize(1));
```

**Comment by caleblloyd** [comment]
Fixed in #407
